Datalab local

gramster opened the pull request:
Local execution with user accounts.

qimingj commented on sources/lib/datalab/gcp/datalab/_projects.py:
```diff
@@ -0,0 +1,56 @@
+  return _utils.handle_magic_line(line, cell, parser)
+
+
+def _list_line(args, _):
```
Why not _list_projects?

gramster replied:
Because it handles the list subcommand of the projects magic, and it is a line magic, not a cell magic (else it would have been called _list_cell).

qimingj commented on containers/datalab/Dockerfile.in:
```diff
@@ -22,112 +22,126 @@ RUN echo "deb-src http://ftp.us.debian.org/debian testing main" >> /etc/apt/sour
+    wget https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.7.1-cp27-none-linux_x86_64.whl && \
+    pip install tensorflow-0.7.1-cp27-none-linux_x86_64.whl && \
+    rm tensorflow-0.7.1-cp27-none-linux_x86_64.whl && \
+    easy_install https://github.com/GoogleCloudPlatform/DataflowPythonSDK/archive/v0.2.2.tar.gz && \
```
This is an internal version of Dataflow that I would hesitate the send it out, especially with the following tweaks in /usr/local/lib/python2.7/dist-packages/google/__init__.py.

gramster replied:
Removed

qimingj commented on containers/datalab/content/run-debug.sh:
```diff
@@ -15,21 +15,25 @@
 # Runs the docker container locally.
 
-export DATALAB_ENV=debug
+# We use the presence of /root/.config to tell whether to show landing page.
```
As I know run-debug.sh is a mode between local managed VM run and real managed VM run, which is supposed to debugging managed VM Datalab. We probably don't need it here.

gramster replied:
Repurposed for node debugging.

qimingj commented on containers/datalab/debug.sh:
```diff
@@ -18,33 +18,24 @@
 ENTRYPOINT="/datalab/run-debug.sh"
-if [ "$1" == "shell" ]; then
-  ENTRYPOINT="/bin/bash"
+if [ "$1" != "" ]; then
```
Remove this file too?

gramster replied:
No, we may still want to do this. E.g. if we run the debugger inside the container.

qimingj commented on containers/datalab/Dockerfile.in:
```diff
@@ -22,112 +22,126 @@ RUN echo "deb-src http://ftp.us.debian.org/debian testing main" >> /etc/apt/sour
 # Container configuration
 EXPOSE 8080
-VOLUME [ "/var/log/app_engine", "/nb", "/root/.config/gcloud" ]
+VOLUME [ "/notebooks" ]
```
Not sure about this. Do we still need /notebooks?

gramster replied:
That's where the user maps in their notebooks.

gramster commented:
Review status: 0 of 130 files reviewed at latest revision, 5 unresolved discussions.

---

*[containers/datalab/debug.sh, line 23 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0Y-r1-23:-KFja_Lr9B4wtOTQ3NDc:-1022585975) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/debug.sh#L23)):*
Added a comment about it being for enabling the node debugger.

---

*[containers/datalab/Dockerfile.in, line 90 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0S-r1-90:-KFjaowXptOHEJK0ug2X:436562334) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/Dockerfile.in#L90)):*
We need some location for users to map in the external directory they are going to use.

---

*[containers/datalab/Dockerfile.in, line 120 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0S-r1-120:-KFjcjjZ15F_aTyuy5UE:610327451) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/Dockerfile.in#L120)):*
Removed

---

*[containers/datalab/content/run-debug.sh, line 18 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0V-r1-18:-KFjco9VLEdJY6lVCq2n:1346508828) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/content/run-debug.sh#L18)):*
I repurposed it to enable node debugging.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799)*
<!-- Sent from Reviewable.io -->

gramster commented:
Review status: 0 of 130 files reviewed at latest revision, 5 unresolved discussions.

---

*[containers/datalab/debug.sh, line 23 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0Y-r1-23:-KFjdOpb5qOfKtk0k89i:-499146009) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/debug.sh#L23)):*
Done.

---

*[containers/datalab/Dockerfile.in, line 90 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0S-r1-90:-KFjdR1CYuqvnLR6qgnW:-499146009) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/Dockerfile.in#L90)):*
Done.

---

*[containers/datalab/Dockerfile.in, line 120 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0S-r1-120:-KFjdS7mj_xXjwEwNYmt:-499146009) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/Dockerfile.in#L120)):*
Done.

---

*[containers/datalab/content/run-debug.sh, line 18 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzWK-mutQ8xhL0V-r1-18:-KFjdTBRtQppkAC3ur1W:-499146009) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/containers/datalab/content/run-debug.sh#L18)):*
Done.

---

*[sources/lib/datalab/gcp/datalab/_projects.py, line 47 \[r1\]](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799#-KFj_FzaaiaOVLXNoI3K-r1-47:-KFjdTyl1Gdi3Ro1IHYz:-499146009) ([raw file](https://github.com/googlecloudplatform/datalab/blob/d0a08c64bf9a0e596c9eec0b65ed2480a4c36251/sources/lib/datalab/gcp/datalab/_projects.py#L47)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/datalab/799)*
<!-- Sent from Reviewable.io -->

qimingj commented on sources/lib/api/gcp/context/_context.py:
```diff
@@ -52,22 +58,25 @@ def project_id(self):
     The default Context is based on project id and credentials inferred from
     metadata returned by the cloud metadata service. It is also managed as a
-    global shared instance used everytime the default context is retrieved.
+    global shared instance used every time the default context is retrieved.
```
The metadata is no longer needed. Update comments?

gramster replied:
Done

qimingj commented on sources/web/datalab/server.ts:
```diff
@@ -173,55 +169,111 @@ function handledAuthenticatedRequest(request: http.ServerRequest,
+        } else {
+          // Push them to Jupyter and handle request.
+          oauth2Client.setCredentials(credentials);
+          persistCredentials(credentials);
```
The TODO needs to be done. :) Regardless of whether refreshAccessToken actually refreshes the token each time, we can't write to disk for each request.

qimingj replied:
Since this is a fairly large CL, perhaps we can address it in a separate change but it needs to be done.

gramster replied:
Done

qimingj commented on sources/lib/api/gcp/context/_project.py:
```diff
@@ -0,0 +1,94 @@
+      list_info = self._api.projects_list(max_results=count,
+                                          page_token=page_token)
+    except Exception as e:
+      raise e
```
Remove the try except since you just raise it again?

gramster replied:
We do this in a number of places to reduce the size of the exception stack shown to users.

qimingj commented on sources/lib/datalab/gcp/datalab/__init__.py:
```diff
@@ -86,3 +94,24 @@ def _run_cell_magic(self, magic_name, line, cell):
+
+try:
+  if 'project_id' not in _IPython.get_ipython().user_ns:
+    _IPython.get_ipython().user_ns['project_id'] = _get_project_id
```
Possibly in a separate change, expose it as a param in datalab.sh so it can be used as default? It's tedious to set it for each notebook.

gramster replied:
Done

qimingj commented on sources/lib/datalab/setup.py:
```diff
@@ -10,20 +10,27 @@
+    'ipython>=4.0,<4.1',
+    'pandas>=0.17.1',
+    'pandas-profiling==1.0.0a2',
+    'tensorflow==0.7.1',
```
tensorflow is not required?

gramster replied:
Removed from here and the api lib setup.

qimingj commented on sources/web/datalab/server.ts:
```diff
@@ -173,55 +169,111 @@ function handledAuthenticatedRequest(request: http.ServerRequest,
-      handledAuthenticatedRequest(request, response, path);
-    }
-    else {
+  logging.getLogger().info('REMOTE <' + request.connection.remoteAddress + '>');
```
Do we need logging for each request although it is info? Seems too noisy.

gramster replied:
Sorry, this should have been removed. Gone now.

qimingj commented on sources/web/datalab/server.ts:
```diff
@@ -173,55 +169,111 @@ function handledAuthenticatedRequest(request: http.ServerRequest,
+    } else {
+      // First time auth.
+      var OAuth2: any = google.auth.OAuth2;
+      // TODO(gram): can we get the port from somewhere instead of hard-coding?
```
Not just port, also host name? Hard-coding it to localhost breaks the story of setting it up on a server and browse it from another machine, which might be useful in some cases.

gramster replied:
For now I don't think we can address this easily without knowing what that future hosted version is using. Changed the comment anyway.

qimingj commented:
Done reviewing. A few observations.

gramster commented:
PTAL

qimingj commented:
LGTM.
